src/attach-shadow.js

```diff
@@ -90,94 +94,142 @@ ShadyRoot.prototype._rendererForHost = function() {
+  }
+  if (!this._slotMap) {
+    this._updateSlotMap();
+  }
+  // distribute host children.
+  for (let n=this.host.firstChild; n; n=n.nextSibling) {
+    this._distributeNodeToSlot(n);
+  }
+  // fallback content, slotchange, and dirty roots
+  for (let i=0, slot; i < this._slotList.length; i++) {
+    slot = this._slotList[i];
+    // distribute fallback content
+    if (!slot.__shady.assignedNodes.length) {
+      for (let n=slot.firstChild; n; n=n.nextSibling) {
+        this._distributeNodeToSlot(n, slot);
+      }
+    }
+    let slotParentRoot = slot.parentNode.shadowRoot;
+    if (slotParentRoot && slotParentRoot._hasInsertionPoint()) {
+      slotParentRoot['_renderRoot']();
+    }
+    slot.__shady.distributedNodes = [];
+    this._addDistributedNodes(slot.__shady.distributedNodes, slot);
+    let prevAssignedNodes = slot.__shady._previouslyAssignedNodes;
+    if (prevAssignedNodes) {
+      for (let i=0; i < prevAssignedNodes.length; i++) {
+        prevAssignedNodes[i].__shady._prevAssignedSlot = null;
+      }
+      slot.__shady._previouslyAssignedNodes = null;
+      // dirty if previously less assigned nodes than previously assigned.
+      if (prevAssignedNodes.length > slot.__shady.assignedNodes.length) {
+        slot.__shady.dirty = true;
+      }
+    }
+    if (slot.__shady.dirty) {
+      slot.__shady.dirty = false;
+      this._fireSlotChange(slot);
+    }
+  }
 }
 
-ShadyRoot.prototype.distribute = function() {
-  let dirtyRoots = this._distributor.distribute();
-  for (let i=0; i<dirtyRoots.length; i++) {
-    dirtyRoots[i]['_render']();
+ShadyRoot.prototype._distributeNodeToSlot = function(node, slot) {
+  node.__shady = node.__shady || {};
+  let oldSlot = node.__shady._prevAssignedSlot;
+  node.__shady._prevAssignedSlot = null;
+  if (!slot) {
+    let name = node.slot || CATCHALL_NAME;
+    slot = this._slotMap[name];
+  }
+  if (slot) {
+    slot.__shady.assignedNodes.push(node);
+    node.__shady.assignedSlot = slot;
+  } else {
+    node.__shady.assignedSlot = undefined;
+  }
+  if (oldSlot !== node.__shady.assignedSlot) {
+    // TODO(sorvell): add test to verify this is needed.
+    if (oldSlot) {
+      oldSlot.__shady.dirty = true;
+    }
+    if (node.__shady.assignedSlot) {
+      node.__shady.assignedSlot.__shady.dirty = true;
+    }
   }
 }
 
-ShadyRoot.prototype.updateInsertionPoints = function() {
-  let i$ = this._insertionPoints;
-  // if any insertion points have been removed, clear their distribution info
-  if (i$) {
-    for (let i=0, c; i < i$.length; i++) {
-      c = i$[i];
-      if (c.getRootNode() !== this) {
-        this._distributor.clearAssignedSlots(c);
+ShadyRoot.prototype._clearSlotAssignedNodes = function(slot) {
+  let n$ = slot.__shady.assignedNodes;
+  slot.__shady.assignedNodes = [];
+  slot.__shady._previouslyAssignedNodes = n$;
+  if (n$) {
+    for (let i=0; i < n$.length; i++) {
+      let n = n$[i];
+      n.__shady._prevAssignedSlot = n.__shady.assignedSlot;
```

Comment:
Yes, but `_prevAssignedSlot` is used only to track if `slotchange` should fire. In the noted case it should so this doesn't seem like an issue? I'll add a comment to note it.